pir/pir-internal/src/main/java/com/duckduckgo/pir/internal/scan/PirScan.kt

```diff
@@ -0,0 +1,274 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.pir.internal.scan
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.webkit.WebView
+import com.duckduckgo.common.utils.DispatcherProvider
+import com.duckduckgo.di.scopes.ServiceScope
+import com.duckduckgo.pir.internal.component.PirDetachedWebViewProvider
+import com.duckduckgo.pir.internal.scripts.BrokerActionProcessor
+import com.duckduckgo.pir.internal.scripts.BrokerActionProcessor.ActionResultListener
+import com.duckduckgo.pir.internal.scripts.PirCssScriptLoader
+import com.duckduckgo.pir.internal.scripts.models.BrokerAction
+import com.duckduckgo.pir.internal.scripts.models.BrokerAction.Navigate
+import com.duckduckgo.pir.internal.scripts.models.PirErrorReponse
+import com.duckduckgo.pir.internal.scripts.models.PirSuccessResponse
+import com.duckduckgo.pir.internal.scripts.models.PirSuccessResponse.ExtractedResponse
+import com.duckduckgo.pir.internal.scripts.models.PirSuccessResponse.NavigateResponse
+import com.duckduckgo.pir.internal.scripts.models.ProfileQuery
+import com.duckduckgo.pir.internal.scripts.models.asActionType
+import com.duckduckgo.pir.internal.store.PirRepository
+import com.squareup.anvil.annotations.ContributesBinding
+import javax.inject.Inject
+import kotlinx.coroutines.runBlocking
+import logcat.logcat
+
+interface PirScan {
+    /**
+     * This method can be used to execute pir scan for a given list of [brokers] names.
+     *
+     * @param brokers List of broker names
+     * @param context Context in which we want to create the detached WebView from
+     * @param onScanComplete callback method called when the scan is completed.
+     */
+    fun execute(
+        brokers: List<String>,
+        context: Context,
+        onScanComplete: () -> Unit,
+    )
+
+    /**
+     * This method can be used to execute pir scan for all active brokers (from json).
+     *
+     * @param context Context in which we want to create the detached WebView from
+     * @param onScanComplete callback method called when the scan is completed.
+     */
+    fun executeAllBrokers(
+        context: Context,
+        onScanComplete: () -> Unit,
+    )
+
+    /**
+     * This method takes care of stopping the scan and cleaning up resources used.
+     */
+    fun stop()
+}
+
+@ContributesBinding(
+    scope = ServiceScope::class,
+    boundType = PirScan::class,
+)
+@ContributesBinding(
+    scope = ServiceScope::class,
+    boundType = ActionResultListener::class,
+)
```

Comment:
If we don't add the `SingleInstanceIn` annotation I guess we'll be creating two instances of `RealPirScan`, one for `PirScan` and another one for the `ActionResultListener` type. Is that the intended behavior?